crates/file-serve/src/lib.rs

```diff
@@ -100,17 +105,54 @@ impl Server {
         self.addr.as_str()
     }
 
+    /// Whether the server was running at the instant the call happened
+    pub fn is_running(&self) -> bool {
+        self.running.load(Ordering::Acquire)
+    }
+
     /// Start the webserver
-    pub fn serve(&self) -> Result<std::convert::Infallible, Error> {
-        // attempts to create a server
-        let server = tiny_http::Server::http(self.addr()).map_err(Error::new)?;
+    pub fn serve(&self) -> Result<(), Error> {
+        if self
+            .running
+            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
+            .is_err()
+        {
+            return Err(Error::new("the server is running"));
+        }
 
-        for request in server.incoming_requests() {
-            if let Err(e) = static_file_handler(self.source(), request) {
-                log::error!("{}", e);
+        // attempts to create a server
+        let server = match tiny_http::Server::http(self.addr()) {
+            Ok(server) => server,
+            Err(e) => {
+                self.running.store(false, Ordering::Release);
+                return Err(Error::new(e));
+            }
+        };
+
+        while self.running.load(Ordering::Acquire) {
+            match server.try_recv() {
```

Comment:
This is changing `serve` from letting the process sleep between invocations (I presume) to being in a busy-wait.

We could put sleeps in here but that just adds latency while still being in a slightly less bad busy-wait.

tiny_http has [unblock](https://docs.rs/tiny_http/latest/tiny_http/struct.Server.html#method.unblock)  to help with cancellations though that will likely require a bit bigger of an API change to support.

*(sorry, somehow overlooked this in the first pass)*